crates/bevy_ecs/src/schedule/state.rs

```diff
@@ -1,260 +1,535 @@
 use crate::{
     component::Component,
-    schedule::{Stage, SystemDescriptor, SystemStage},
-    world::World,
+    schedule::{ShouldRun, SystemSet},
+    system::{In, IntoChainSystem, IntoSystem, Local, Res, ResMut, System},
 };
-use bevy_utils::HashMap;
-use std::{mem::Discriminant, ops::Deref};
 use thiserror::Error;
 
-pub(crate) struct StateStages {
-    update: Box<dyn Stage>,
-    enter: Box<dyn Stage>,
-    exit: Box<dyn Stage>,
-}
-
-impl Default for StateStages {
-    fn default() -> Self {
-        Self {
-            enter: Box::new(SystemStage::parallel()),
-            update: Box::new(SystemStage::parallel()),
-            exit: Box::new(SystemStage::parallel()),
-        }
-    }
+/// ### Stack based state machine
+///
+/// This state machine has three operations: Push, Pop, and Next.
+/// * Push pushes a new state to the state stack, pausing the previous state
+/// * Pop removes the current state, and unpauses the last paused state.
+/// * Next unwinds the state stack, and replaces the entire stack with a single new state
+#[derive(Debug)]
+pub struct State<T: Component + Clone + Eq> {
+    transition: Option<StateTransition<T>>,
+    stack: Vec<T>,
+    scheduled: Option<ScheduledOperation<T>>,
+    ran_update: bool,
 }
 
-pub struct StateStage<T> {
-    stages: HashMap<Discriminant<T>, StateStages>,
+#[derive(Debug)]
+enum StateTransition<T: Component + Clone + Eq> {
+    PreStartup,
+    Startup,
+    // The parameter order is always (leaving, entering)
+    ExitingToResume(T, T),
+    ExitingFull(T, T /**/),
```

Comment:
Why the `/**/`?

Reply:
It was an internal marker I forgot to remove